### PR TITLE
Fix issues around MailboxProcessor

### PR DESCRIPTION
**Pull request by forki**
related to https://github.com/Microsoft/visualfsharp/pull/3369 - but fix is much harder because races ...

### CODE REVIEW

**Comment by forki** [comment]
It looks like ubuntu build is very flaky

**Comment by forki** [comment]
includes https://github.com/Microsoft/visualfsharp/pull/3369

File: src/fsharp/FSharp.Core/control.fs [hunk 1]
```diff
@@ -2437,30 +2421,34 @@ namespace Microsoft.FSharp.Control
...
-                    |   None -> 
-                            let! ct = Async.CancellationToken
-                            let timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct, CancellationToken.None)
-                            let timeoutAsync = AsyncHelpers.timeout timeout timeoutCts.Token
-                            return! scan timeoutAsync timeoutCts
-                    |   Some resP -> 
-                            let! res = resP
-                            return Some res
-
+            async {
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: Please make this diff absolutely minimal, e.g. avoid any reformatting of code unless totally necessary, thanks

Comment (forki) [inline src/fsharp/FSharp.Core/control.fs]: I think @matthid said rewrite would be necessary. :-)

Comment (forki) [inline src/fsharp/FSharp.Core/control.fs]: ok restored formatting

**Comment by dsyme** [comment]
@forki Which bug is this fixing? thanks

File: src/fsharp/FSharp.Core/control.fs [hunk 2]
```diff
@@ -2304,37 +2304,21 @@ namespace Microsoft.FSharp.Control
         let ensurePulse() = 
             match pulse with 
             | null -> 
-                pulse <- new AutoResetEvent(false);
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: remove formatting change for this fix, thanks

**Comment by forki** [comment]
#3337 and analogous cases in scan

File: src/fsharp/FSharp.Core/control.fs (hunk 2 above)
Comment (forki) [inline src/fsharp/FSharp.Core/control.fs]: restored that line. that's my first semicolon in years ;-)

**Comment by dsyme** [comment]
@forki OK, thanks

**Comment by dsyme** [comment]
@forki As I understand it the primary bug is that using ``Receive`` with no timeout is not cancellable when an explicit cancellation token has been provided.  (The workaround to the bug is to use a really large timeout, which activates cancellation)

Looking at your code, I assume you've concluded that it's not possible (or very difficult) to use the ``waitOneNoTimeout`` code, because of the possibility that a cancellation token occurs.  

However, with your change ``savedCont`` is no longer being set in any path.  We will need to look at this closely, as my understanding is that the ``savedCont`` hack is an important optimization for the common case where there is no cancellation and no timeout.

It seems that at [this line](https://github.com/Microsoft/visualfsharp/pull/3370/files#diff-34ee82347e9e0b47265fcecaa4850174R2505) we can detect the common case where no cancellation token has been provided. Although we default to using the global ``Async.DefaultCancellationToken``, and in theory cancellation may be activated on that token, that behaviour is not critical and I would be happy to accept a fix where ``Receive`` with no timeout was only cancellable on an __explicit__ cancellation token on the constructor to the MailboxProcessor type (as given in the repro to #3337).

So do you think it would be possible to change the fix so that the existing ``waitOneNoTimeout`` code is used when no explicit cancellation token has been given on the constructor to the MailboxProcessor type?  I think you would [pass in a boolean to the constructor of ``Mailbox``](https://github.com/Microsoft/visualfsharp/pull/3370/files#diff-34ee82347e9e0b47265fcecaa4850174R2506) indicating whether cancellability was being activated by the presence of a non-default ``cancellationToken`` arg

If we did remove the ``savedCont`` hack then we would need to check the performance of basic MailboxProcessor code like this
```
let counter =
    MailboxProcessor.Start(fun inbox -> async {
        while true do 
             let! msg = inbox.Receive()
             count <- count + msg 
       })
```

**Comment by forki** [comment]
tbh I don't think the fixes are really good. I think I will revert the fixes and only contribute the broken tests. this is too dangerous area for me.

**Comment by dsyme** [comment]
@forki I updated the PR with the minimal fix as I can see it - could you take a look?  thanks

**Comment by forki** [comment]
@matthid what do you think?

**Comment by matthid** [comment]
It now still has the race condition I have shown when using a timeout. Which means the timeout will be triggered while a message was already available all the waiting time...

To be fair it is not a regression, but as we already identified it we probably want to fix it here

**Comment by forki** [comment]
But the test is green, right? Or did I forget one?

Am 26.07.2017 8:28 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by matthid** [comment]
Copy the test and use the timeout overload, however it depends a bit on what recieve is doing on a timeout and might need a slight modification. Is it throwing on timeout?

**Comment by forki** [comment]
can you PR it? Is probably faster than explaining to me ;-)

**Comment by dsyme** [comment]
Yes, please feel free to add a failing test to this PR if there is another case to fix. Or create a separate issue with reliable repro

**Comment by forki** [comment]
@matthid is this ready to go?

**Comment by matthid** [comment]
Sorry forgot about this. As I said the race is still obvious (from looking at the code). 

But it is not a regression for the timeout api. I'll leave it up to the visualfsharp maintainers how they feel about this. You can argue either way, however I'd personally not merge it according to clean code philosophy (pfadfinderregel)

**Comment by forki** [comment]
Could you please send a failing test then?

Am 07.08.2017 11:41 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by matthid** [comment]
Ok done at https://github.com/forki/visualfsharp/pull/29.
@dsyme can you take a look or suggest how we should fix it (ie re-add the lock-check from @forki ?)

**Comment by matthid** [comment]
Haha, the test was green locally as well the first time. But it definitely failed consistently afterwards. Maybe I should have added another "0" to the number of retries. 

Can someone restart the CI?

**Comment by dsyme** [comment]
@matthid Thanks!  Could you describe the race?  I took a look through the code and believe I saw one around the creation of ``pulse`` but perhaps there is another too

**Comment by matthid** [comment]
@dsyme First: I don't know why the CI is green, locally it fails consistently

Second:
Yes the race is around pulse:

https://github.com/Microsoft/visualfsharp/blob/f535de60f709f25485d6dc97a9639b53a345bd40/src/fsharp/FSharp.Core/control.fs#L2389-L2405

https://github.com/Microsoft/visualfsharp/blob/f535de60f709f25485d6dc97a9639b53a345bd40/src/fsharp/FSharp.Core/control.fs#L2304-L2310


If you look at the code it should be straightforward to see that the thread posting the arrival could be on line 2398 before the receiving one finished 2307. In that case the receiving one will wait for a message when it actually is already in the arrivals list.

This code looks so dangerous and error prone I probably shouldn't even try to suggest a fix here (besides rewriting the whole thing - obviously), but we could re-check the arrivals list after registering the pulse?

**Comment by dsyme** [comment]
@matthid Thanks. I believe we just lock the access to pulse with ``syncObj``, e.g. this:

         let ensurePulse() =  
             match pulse with  
             | null ->  
                 lock syncObj (fun () -> 
                     match pulse with  
                     | null ->  pulse <- new AutoResetEvent(false); 
                     | _ -> ())
             | _ ->  
                 () 
             pulse 

The other case is already locked by ``syncObj``

**Comment by matthid** [comment]
@dsyme How does that help?

**Comment by dsyme** [comment]
The race you mentioned was this:

> ...the thread posting the arrival could be on line 2398 before the receiving one finished 2307. 

* 2398 is and was `` match pulse with  | null ->  `` inside a ``lock syncRoot``
* 2307 is now   ``pulse <- new AutoResetEvent(false); `` inside a ``lock syncRoot``

**Comment by matthid** [comment]
Wait, maybe I don't get it but with your "fix" it basically still does the wrong thing but slower:

T1: L2398 -> arrivals was modified
T2: Waiting for lock
T1: Finishes the lock by doing nothing (as pulse was null)
T2: Continues with the wrong behavior (setting the pulse object and waiting on it).

**Comment by matthid** [comment]
Note that this issue means in practice that messages could arrive in the wrong order, as another incoming message would probably "heal" the situation. I'm mentioning this as we might have reports open?

**Comment by dsyme** [comment]
@matthid OK, I understand what you're referring to now, thanks.  Will work on proposing a fix

**Comment by dsyme** [comment]
@matthid I pushed a proposed fix - basically to force the creation of the pulse when a non-infinite-timeout or cancellable Receive/TryReceive is performed - and then it goes back and checks the arrivals again. Please take a look

Running locally the new test passes, but let's see what CI says.

**Comment by dsyme** [comment]
> @dsyme First: I don't know why the CI is green, locally it fails consistently

Could you check the failure on your machine please?  Thanks

**Comment by matthid** [comment]
@dsyme Yes, the fix seems to work. And it looks fine from the assumption that only a single thread is ever receiving (and I never saw anyone using this differently).

A bit funny how you re-added @forki's cleanup ;). Sometimes it feels a bit like people in this repository only want to irritate "outsiders". But it's fine - no reason to open such a big discussion here. I'm really glad for every cleanup this repository gets (even if it's hard to get it in from the outside) :)

Thanks for taking care of this.

**Comment by dsyme** [comment]
> A bit funny how you re-added @forki's cleanup ;). 

I'll take it out.  There are very good reasons for minimizing diffs, **especially** ones that fix race conditions, and you are right to point it out.  No one is trying to irritate anyone, we should all be vigilant in enforcing high standards of PRs that will help document the change history.

**Comment by dsyme** [comment]
@matthid Done, thanks.  BTW the main reasons to minimize a diff are

- We (Microsoft) sometimes need to cherry-pick these , e.g. to prepare a hotfix in a customer deployment. That's not at all common, but for something like this it's not entirely out of the question

- We (Microsoft) also sometimes have these fixes reviewed by non-F# experts ("war rooms") when integrating them into otherwise-locked-down release branches

- We need to look back at diffs to understand the history of changes when fixing later bugs or making changes

In all cases, minimal fixes can help.  They aren't essential all the time though, but for this kind of thing they can be important

**Comment by dsyme** [comment]
@matthid Thanks a million for validating the fix.  It would be great to get the race checked by CI as well, I suppose it is VM providing different timings

**Comment by dsyme** [comment]
@KevinRansom this is ready

**Comment by forki** [comment]
Can you please apply the cleanup afterwards? ;-)

Am 18.08.2017 16:53 schrieb "Don Syme" <notifications@github.com>:

**Comment by matthid** [comment]
@dsyme Yes thanks for the in-depth explanation. It's completely fine with me. I was just pointing out the irritating fact. In fact I would even understand having "higher" bars for external contributors if it is transparent in what they are and where/how they apply.

**Comment by dsyme** [comment]
> @dsyme Yes thanks for the in-depth explanation. It's completely fine with me. I was just pointing out the irritating fact. 

Yes, though instead of getting irritated, please police us as if we were equal contributors.  In the long run that's better for all.

I personally probably get scrutinized less than others in this repo, since people often assume I know what I'm doing.  But that's not always the case.  I'm glad you're scrutinizing my work in particular, and others should do the same.

**Comment by forki** [comment]
Don't worry. I learned very early in paket and FAKE that @matthid is very good at playing this game. :P

**Comment by forki** [comment]
@kevinransom this is ready

**Comment by dsyme** [comment]
@matthid Could you take a look at https://github.com/Microsoft/visualfsharp/pull/3498 for me please?  In the common case where pulse stays null (timeout = -1, no cancellation) we have an endless async loop

**Pull request merged**